solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/SolaceSessionManager.java

```diff
@@ -0,0 +1,158 @@
+package com.solace.spring.cloud.stream.binder;
+
+import static com.solacesystems.jcsmp.XMLMessage.Outcome.ACCEPTED;
+import static com.solacesystems.jcsmp.XMLMessage.Outcome.FAILED;
+import static com.solacesystems.jcsmp.XMLMessage.Outcome.REJECTED;
+
+import com.solace.spring.cloud.stream.binder.health.handlers.SolaceSessionEventHandler;
+import com.solacesystems.jcsmp.Context;
+import com.solacesystems.jcsmp.ContextProperties;
+import com.solacesystems.jcsmp.JCSMPException;
+import com.solacesystems.jcsmp.JCSMPProperties;
+import com.solacesystems.jcsmp.JCSMPSession;
+import com.solacesystems.jcsmp.SolaceSessionOAuth2TokenProvider;
+import com.solacesystems.jcsmp.SpringJCSMPFactory;
+import com.solacesystems.jcsmp.impl.JCSMPBasicSession;
+import com.solacesystems.jcsmp.impl.client.ClientInfoProvider;
+import java.util.Set;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Value;
+import org.springframework.lang.Nullable;
+
+public class SolaceSessionManager implements SolaceSessionProvider {
+
+  private static final Logger LOGGER = LoggerFactory.getLogger(SolaceSessionManager.class);
+
+  private final JCSMPProperties jcsmpProperties;
+  private final ClientInfoProvider clientInfoProvider;
+  private final SolaceSessionEventHandler solaceSessionEventHandler;
+  private final SolaceSessionOAuth2TokenProvider solaceSessionOAuth2TokenProvider;
+  private final boolean failOnStartup;
+
+  private volatile JCSMPSession jcsmpSession;
+  private volatile Context context;
+  private final Object sessionLock = new Object();
+
+  public SolaceSessionManager(JCSMPProperties jcsmpProperties,
+      @Nullable ClientInfoProvider clientInfoProvider,
+      @Nullable SolaceSessionEventHandler eventHandler,
+      @Nullable SolaceSessionOAuth2TokenProvider solaceSessionOAuth2TokenProvider,
+      @Value("${solace.session.failOnStartup:true}") boolean failOnStartup) {
```

Comment:
`fail-on-startup` isn't really accurate here I think. This more like a `lazy-connect` feature. So an enumerated property like `connection-timing=(eager|lazy)` might be better